doc/source/development/rolling_operations_with_numba.rst

```diff
@@ -0,0 +1,173 @@
+| mean (fixed window)     | 37.2 ms ± 1.66 ms| 21 ms ± 1.03 ms |
++-------------------------+------------------+-----------------+
+| mean (offset window)    | 102 ms ± 2.73 ms | 28.2 ms ± 735 µs|
++-------------------------+------------------+-----------------+
+| apply (fixed window)    | 577 ms ± 8.95 ms | 3.29 s ± 117 ms |
++-------------------------+------------------+-----------------+
+| apply (offset window)   | 640 ms ± 5.51 ms | 3.54 s ± 98.6 ms|
++-------------------------+------------------+-----------------+
+
++-------------------------+------------------+-----------------+
+| Peak Memory             | Numba            | Cython          |
++=========================+==================+=================+
+| mean (fixed window)     | 240.22 MiB       | 161.13 MiB      |
++-------------------------+------------------+-----------------+
+| mean (offset window)    | 245.81 MiB       | 177.08 MiB      |
++-------------------------+------------------+-----------------+
+| apply (fixed window)    | 268.28 MiB       | 177.12 MiB      |
++-------------------------+------------------+-----------------+
+| apply (offset window)   | 277.91 MiB       | 184.79 MiB      |
++-------------------------+------------------+-----------------+
+
+Although peak memory has consistently increased and there is a slight performance decrease for rolling mean,
+rolling apply benchmarks have greatly increased.
+
+Window Indexer API
+------------------
+
+A ``BaseIndexer`` class is available from ``pandas/core/window/indexers.py`` for users to subclass
+to create a custom routine to calculate window boundaries. Users will need to specify a
+``get_window_bounds`` function to calculate window boundaries. Below is an example of creating an
+``ExpandingIndexer`` computes an expanding window:
+
+.. code-block:: python
+
+   from pandas import Series
+   from pandas.core.window.indexers import BaseIndexer
+
+   class ExpandingIndexer(BaseIndexer):
+       """Calculate expanding window bounds."""
+
+       def get_window_bounds(
+           self,
+           num_values: int = 0,
+           window_size: int = 0,
+           min_periods: Optional[int] = None,
+           center: Optional[bool] = None,
+           closed: Optional[str] = None,
+           win_type: Optional[str] = None,
+       ) -> BeginEnd:
+           """
+           Examples
+           --------
+           >>> ExpandingIndexer().get_window_bounds(10)
+
+           (array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+            array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10]))
+           """
+           return np.zeros(num_values, dtype=np.int64), np.arange(1, num_values + 1)
+
+   s = Series(range(10))
+   s.rolling(ExpandingIndexer()).mean()
+
+
+Appendix
+--------
+
+Timings on master:
+
+.. code-block:: ipython
+
+   In [1]: %load_ext memory_profiler
+
+   In [2]: pd.__version__
+   Out[2]: '0.26.0.dev0+514.g24b1dd128'
+
+   In [3]: n = 1_000_000
```

Comment:
Let's try these with 10M, 100M and 1B

Reply:
For memory, which we were most curious about, at 1B the Numba implimentation is finally starting to edge out Cython. Does that align with your experience?

```
# POC
In [1]: %load_ext memory_profiler

In [2]: n = 1_000_000_000

In [3]: roll_fixed = pd.Series(range(n)).rolling(10)

In [4]: %memit roll_fixed.apply(lambda x: np.sum(x) + 5, raw=True)
peak memory: 11747.92 MiB, increment: 4056.36 MiB

# master
In [1]: %load_ext memory_profiler

In [2]: n = 1_000_000_000

In [3]: roll_fixed = pd.Series(range(n)).rolling(10)

In [4]: %memit roll_fixed.apply(lambda x: np.sum(x) + 5, raw=True)
peak memory: 12581.71 MiB, increment: 4860.52 MiB
```

Reply:
Interesting. I think it is possible that rolling in Cython is just pretty efficient. The time I got a significant memory improvement over Pandas was comparing with a Numba EWA implementation, which is definitely more complex.

Reply:
I think it is worth mentioning somewhere in the doc that we have already found that, as N increases, the memory difference decreases and eventually Numba beats the existing impl.

Reply:
Sounds good. Noted the long term memory pattern